clustering.py

```diff
@@ -244,3 +246,220 @@ def sort_labels(label_dict, cluster_labels, ts_to_labels):
         ordered_labels[i] = system_labels[elt]
 
     return ordered_labels, ordered_cluster_labels, ordered_ts_labels
+
+def kmeans_constrained(data, label_dict, ts_to_labels):
+    """Runs k-means with constraints and uses k-means++ initialization.
+
+    Args:
+        data: An np array where each row is a time series and each
+            column is a time.
+        label_dict: A dictionary where the keys are labels and the
+            values are the indexes of the labels in data.
+        ts_to_labels: An array where each row is a timeSeries and each
+            column is a label.
+
+    Returns:
+        An np array where the ith element is the cluster the ith time
+        series was placed in.
+    """
+    must_link, can_not_link = make_constraints(ts_to_labels)
+    num_clusters = (len(data) // 20) + 6
+    clusters_distances = []
+
+    for run_num in range(5):
```

Comment:
For k-means it is common to reinitialize the centroids and run k-means multiple times since some centroids may work better than others. I will update it to 10, which is more common than 5.